src/lib/core/RoomConnection/redux/listenerMiddleware.ts

```diff
@@ -11,3 +11,29 @@ export type AppStartListening = TypedStartListening<RootState, AppDispatch, Retu
 export const startAppListening = listenerMiddleware.startListening as AppStartListening;
 
 export const addAppListener = addListener as TypedAddListener<RootState, AppDispatch>;
+
+/**
+ * Creates a reactor that will be called on every action.
+ * Every reactor needs to update a piece of state that it depends on, to avoid infinite loops.
+ * example:
+ * ```ts
+ * createReactor((action, { dispatch, getState }) => {
+ *   const stateField = selectState(getState());
+ *
+ *   if (stateField !== "some value") {
+ *      // The action here needs to update the stateField to "some value"
+ *      dispatch(someAction());
+ *   }
+ * }));
+ * ```
+ * @param callback. The callback to be called on every action.
+ * @returns The unsubscribe function.
+ */
+export const createReactor = (
+    callback: ListenerEffect<AnyAction, RootState, AppDispatch, ReturnType<typeof createServices>>
+) => {
+    return startAppListening({
+        predicate: () => true,
```

Comment:
maybe I'm not understanding, but couldn't we prevent running every reactor for every action by moving the `if` logic from the effect to the predicate?

I.e. for the `deviceCredentials` reactor that calls `dispatch(doGetDeviceCredentials()`, we'd want the app listener to look like:
```typescript
startAppListening({
    predicate: () => {
        const wantsToJoin = selectAppWantsToJoin(getState());
        const deviceCredentials = selectDeviceCredentialsRaw(getState());
        return shouldFetchDeviceCredentials(wantsToJoin, deviceCredentials.isFetching, deviceCredentials.data)
    },
    effect: () =>  dispatch(doGetDeviceCredentials()),
})
```

Reply:
or even make `shouldFetchDeviceCredentials` a selector like:
```typescript
const shouldFetchDeviceCredentials = useSelector(selectAppWantsToJoin, selectDeviceCredentialsRaw, (wantsToJoin, deviceCredentials) => {
    if (wantsToJoin && !deviceCredentialsData && !isFetching) {
        return true;
    }
    return false;
}
```

so the reactor call could be simplified to:

```typescript
startAppListening({
    predicate: shouldFetchDeviceCredentials,
    effect: () =>  dispatch(doGetDeviceCredentials()),
})
```

but again, based on what Håvard said in standup I probably don't know the full story 😄